src/pcre2_compile.c

```diff
@@ -5178,7 +5187,722 @@ errorcode = ERR79;
+              } else if (meta == META_KET) {
+                /* Finish last part of group */
+                rewrite_buf_append(buf, META_KET, memctl);
+                goto FINISH_REWRITE;
+              } else {
+                rewrite_buf_copy(buf, copy_from, item_end - copy_from, memctl);
+              }
+              copy_from = item_end;
+            }
+          }
+
+          PCRE2_UNREACHABLE();
+        }
+
+        FINISH_REWRITE:
+        if (quantifier != NULL) {
+          rewrite_buf_copy(buf, quantifier-1, end-quantifier+1, memctl);
+        } else if (copy_opening_paren) {
+          rewrite_buf_append(buf, META_KET, memctl);
+        }
+
+        if (alt_positions != alt_position_buf)
+          memctl->free(alt_positions, memctl->memory_data);
+        return;
+      }
+    }
+
+    if (alt_positions != alt_position_buf)
+      memctl->free(alt_positions, memctl->memory_data);
+  }
+
+  /* We didn't rewrite this group
+   * Even so, a rewrite of some subgroups might be needed */
+  DONT_REWRITE: ;
+  uint32_t *item = first;
+  if (buf->start != NULL) {
+    /* It has already been decided that some part of the overall pattern does need to be rewritten */
+    rewrite_buf_copy(buf, start, first - start, memctl);
+    while (item < end) {
+      uint32_t *item_quant = NULL;
+      uint32_t *item_end = scan_item(item, end, &item_quant);
+      uint32_t meta = META_CODE(*item);
+      if (is_group_starter(meta)) {
+        rewrite_alternation(item, item_end, item_quant, buf, pattern, patstring, memctl);
+      } else {
+        rewrite_buf_copy(buf, item, item_end - item, memctl);
+      }
+      item = item_end;
+      REWRITE_UNDER_WAY: ;
+    }
+  } else {
+    /* It hasn't yet been decided that we need to rewrite some part of the overall pattern...
+     * But if we find a subgroup which needs rewrite, then we can still initiate that process */
+    while (item < end) {
+      uint32_t *item_quant = NULL;
+      uint32_t *item_end = scan_item(item, end, &item_quant);
+      uint32_t meta = META_CODE(*item);
+      if (is_group_starter(meta)) {
+        rewrite_alternation(item, item_end, item_quant, buf, pattern, patstring, memctl);
+        if (buf->start != NULL) {
+          /* A rewrite has now started, so go to the other loop */
+          item = item_end;
+          goto REWRITE_UNDER_WAY;
+        }
+      }
+      item = item_end;
+    }
+  }
+}
+
+static void rewrite_regex(compile_block *cb, pcre2_compile_context *ccontext, BOOL heap_parsed_pattern)
+{
+  rewrite_buf buf;
+  rewrite_buf_init(&buf);
+
+  rewrite_alternation(cb->parsed_pattern, cb->parsed_pattern_end-1, NULL, &buf, cb->parsed_pattern, cb->start_pattern, &ccontext->memctl);
+  rewrite_finish(&buf, cb, &ccontext->memctl, heap_parsed_pattern);
+}
+
+/* Strip non-capturing parentheses which may be wrapping the entire regex */
+static void strip_enclosing_nocaptures(compile_block *cb)
```

Comment:
this is a fairly unique "optimization", care to explain the rationale of why is it needed/desired in the comment instead of the "what" which can be inferred by the code?

Reply:
Will do. The point is basically that the new code adds an extra `(?:` and `)` wrapping every pattern at the top level, in order to make recursive traversal of the pattern easier. So that those do not affect anything in the following stages, it strips them off after pattern rewriting is complete.

Reply:
got it, so the comment should be using "are" instead of "may be" and that also explains the "magic 2" and other oddities like that renamed pointer name and the addition of a clone of it in "intops".

anyway, thanks for doing this and kudos on what looks like a sound first major contribution.

Reply:
As far as I remember all patterns are enclosed in (?: ) by default.

Reply:
@zherczeg You are right, but that is done later in the compilation process. So if I don't strip redundant parens here, then it will mean every pattern will be enclosed in *two* sets of (? :).

Reply:
Added comment here.

Reply:
@alexdowad: I think the argument Zoltan was making was that there is a potential for tighter integration here that is being missed; could `parse_regex()` just add this so your code doesn't add them and then remove them?, could `compile_regex()` be told how to generate the original OP_BRA using those wrapping META instead?

Reply:
Removing unnecessary bracekts are a good idea, and it should improve performance:
```
  re> /a(?s:.)b/B
------------------------------------------------------------------
        Bra
        a
        Bra
        AllAny
        Ket
        b
        Ket
        End
------------------------------------------------------------------
```

The nested Bra / Ket is totally unnecessary. The jit actually optimize them out. However, the interpreter still spends some time to process it, so removing them should be good.

Reply:
> The nested Bra / Ket is totally unnecessary. The jit actually optimize them out. However, the interpreter still spends some time to process it, so removing them should be good.

I guess another option would be to strip out redundant OP_BRA/OP_KET *after* the `parsed_pattern` is converted to bytecode, rather than before.

Anyways, for the moment, I'm happy with this aspect of this PR.